### PR TITLE
refactor: rewrite fluent lib without actors [DET-8303]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
The primary goal of the PR is refactoring the actor-based Fluent daemon manager away from the actor system. Secondary goals, similar to other PRs in this set, are to decouple the interface from other packages (less leaky abstractions) and increase test coverage. There is still one offender w.r.t that: `Options`/`MasterSetAgentOptions` passed everywhere, but I decided this wasn't worth refactoring.

## Test Plan
- [x] has automated tests

## Commentary (optional)
Blocked on #5384 

Order to review this set of PRs:
* [refactor: rewrite websocket lib without actors DET-8299](https://github.com/determined-ai/determined/pull/5382)
* [refactor: rewrite our docker lib without actors DET-8300](https://github.com/determined-ai/determined/pull/4943)
* [refactor: rewrite container lib without actors](https://github.com/determined-ai/determined/pull/5384)
* [refactor: rewrite fluent lib without actors](https://github.com/determined-ai/determined/pull/5385)
* [refactor: rewrite container manager lib without actors](https://github.com/determined-ai/determined/pull/5386)
* [refactor: rewrite agent without actors](https://github.com/determined-ai/determined/pull/5387)

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-8299]: https://determinedai.atlassian.net/browse/DET-8299?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-8300]: https://determinedai.atlassian.net/browse/DET-8300?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: agent/internal/fluent/fluent.go [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+					f.log.Errorf("Fluent Bit: %s", message)
+				}
+			case <-ctx.Done():
+				return fmt.Errorf("fluent bit canceled: %w", ctx.Err())
+			}
+		}
+	})
+	switch err := loggroup.Wait(); {
+	case errors.Is(err, context.Canceled):
+		f.log.WithError(err).Warn("orphaning fluent due to cancellation")
```
Comment (ioga) [inline agent/internal/fluent/fluent.go:178]: why aren't we killing fluent container on cancellation?
isn't cancellation here the normal operation e.g. when the agent process is signaled SIGQUIT, i.e. normal exit?

File: agent/internal/fluent/fluent.go [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+type Fluent struct {
+	// Configuration details.
+	opts  options.AgentOptions
+	mopts aproto.MasterSetAgentOptions
+
+	// System dependencies.
+	log    *log.Entry
+	docker *docker.Client
+
+	// Internal state.
```
Comment (ioga) [inline agent/internal/fluent/fluent.go:56]: `// Circular buffer used for printing recent logs on fluent container failures.`

Comment (ioga) [reply agent/internal/fluent/fluent.go:56]: in fact it seems like this could have been `container/ring` like the one used in container manager

Comment (stoksc) [reply agent/internal/fluent/fluent.go:56]: yeah, didn't change this but i agree, could be. we should just write a generic ring - we already have it in a few places.

File: agent/internal/fluent/fluent.go (hunk 1 above)
Comment (stoksc) [reply agent/internal/fluent/fluent.go:178]: actually, you don't want to kill fluent on cancellation, just on failure you want to make sure you clean up. if you roll the agent to upgrade, you dont want logs to stop working for a short time.

Comment (stoksc) [reply agent/internal/fluent/fluent.go:177]: this logic is replicated on line 186

**Review by ioga — APPROVED** [review]

**Pull request merged**